components/resource_control/src/worker.rs

```diff
@@ -0,0 +1,431 @@
+                return;
+            }
+        };
+        // if total resource quota is unlimited, set all groups' limit to unlimited.
+        if resource_stats.total_quota <= f64::EPSILON {
+            for g in bg_group_stats {
+                limiter_fn(&g.limiter).set_rate_limit(f64::INFINITY);
+            }
+            return;
+        }
+
+        let mut total_ru_quota = 0.0;
+        let mut background_consumed_total = 0.0;
+        let mut has_wait = false;
+        for g in bg_group_stats.iter_mut() {
+            total_ru_quota += g.ru_quota;
+            let total_stats = limiter_fn(&g.limiter).get_statistics();
+            let mut stats_delta =
+                match self.prev_stats_by_group[resource_type as usize].entry(g.name.clone()) {
+                    Entry::Occupied(mut s) => total_stats - s.insert(total_stats),
+                    Entry::Vacant(v) => {
+                        v.insert(total_stats);
+                        total_stats
+                    }
+                };
+            stats_delta = stats_delta / dur_secs;
+            background_consumed_total += stats_delta.total_consumed as f64;
+            g.stats = stats_delta;
+            if stats_delta.total_wait_dur_us > 0 {
+                has_wait = true;
+            }
+        }
+
+        // fast path if process cpu is low
+        let is_low_load = resource_stats.current_used <= (resource_stats.total_quota * 0.1);
+        if is_low_load && !has_wait && self.is_last_time_low_load[resource_type as usize] {
+            return;
+        }
+        self.is_last_time_low_load[resource_type as usize] = is_low_load;
+
+        let mut available_quota = ((resource_stats.total_quota - resource_stats.current_used
+            + background_consumed_total)
+            * 0.9)
+            .max(resource_stats.total_quota * 0.1);
+        let mut total_expected_cost = 0.0;
+        for g in bg_group_stats.iter_mut() {
+            let mut rate_limit = limiter_fn(&g.limiter).get_rate_limit();
+            if rate_limit.is_infinite() {
+                rate_limit = 0.0;
+            }
+            let group_expected_cost = g.stats.total_consumed as f64
+                + g.stats.total_wait_dur_us as f64 / MICROS_PER_SEC * rate_limit;
+            g.expect_cost_per_ru = group_expected_cost / g.ru_quota;
+            total_expected_cost += group_expected_cost;
+        }
+        bg_group_stats.sort_by(|g1, g2| {
+            g1.expect_cost_per_ru
+                .partial_cmp(&g2.expect_cost_per_ru)
+                .unwrap()
+        });
+
+        // quota is enough
+        if total_expected_cost <= available_quota {
+            for g in bg_group_stats.iter().rev() {
+                let expected = g.expect_cost_per_ru * g.ru_quota;
+                let limit = if g.expect_cost_per_ru > available_quota / total_ru_quota {
+                    expected
+                } else {
+                    available_quota / total_ru_quota * g.ru_quota
```

Comment:
I think we can change the `expect_cost_per_ru` to `expect_cost_rate`, and  here maybe :

```
let expected_limit = g.expect_cost_rate (expect_cost_rate)
if expected_limit < avaliable_quota * g.ru_quota / total_ru_quota {
    expected_limit = avaliable_quota * g.ru_quota / total_ru_quota 
}
```

`per_ru` is complicate the logic in here.

and `total_expected_cost <= available_quota` is not always be true in this loop, is it as expect?